existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,264 @@
+Nothing changed since we added the quantifier because it was already there!
+However, there are a few things to notice here.
+
+### Examples
+
+#### Order of type variables
+
+First of all, you can now change the order of type variables.
+Let's imagine you want to define a function with 10 type variables (yeah, there are real-life examples of such functions), and you know that when you use it, you want to specify the last argument explicitly.
+To pass a type as a parameter to a function, we will use the `TypeApplication` extension.
+Without `forall`, you would write something like this:
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+Quite long, right?
+This is because GHC will maintain the same order as when these variables first appear in the function definition.
+However, by using explicit `forall.`, this can be simplified:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
+
+### Support of other extensions
+
+`ExplicitForAll` shines most when you need to enable other extensions – some of them just won't work without the `forall.` quantifier. :)
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to the one defined in a where clause.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = double ++ rest
+  where
+    double :: a -> [a]
+    double x = [x, x]
+```
+This will cause an error without `ScopedTypeVariables`. But to use it, you need to add a `forall.` quantifier.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how it was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+Besides universally quantified variables, Haskell also supports existential types.
+But in order to use them ... you still need to use the `forall.` keyword ¯\\\_(ツ)_/¯.
+
+The following construction is called an existential type in Haskell – when a type variable is introduced (using the `forall.` quantifier) inside the signature, not before it.
+In contrast to the universal quantifier, the existential one means that the compiler (**callee**) is responsible for deducing what a type variable should mean.
+This also means that the caller can't specify this type variable.
+
+![When you use existential types](meme.jpg)
+
+
+### Usage
+
+We will start with functions:
+
+```haskell
+func :: forall a c. a -> (forall b. (Show b) => b) -> c
```

Comment:
Also, I don't think this function is a good example, because it cannot possibly be called.

There's no way you can pass a `(forall b. Show b => b)` to `func`.

Let's take a step back: because `b` is existentially quantified, `func`'s definitinion is gonna responsible for instantiating `b` to some type.

So, at the _use-site_, when you're _using_ func, you have to pass in a function of type `(forall b Show b => b)`. In other words, you have to pass in a function that creates a variable of type `b` no matter what `b` is. That can't be done.

The only way to call this function is to pass in `undefined`.

Reply:
The function you replaced it with (`func :: forall a c. a -> (forall b. (Show b) => b -> IO ()) -> IO c`) can be called, but it _can't be defined_ (without using `undefined`).

Reply:
Well, I would say, that currently this function is some sort of `log` from previous example. What it should do here is a different question. Probably in real, it would be nice to have some constraint CanCast a b for example. But I guess it becomes to complicated for this example

Reply:
This was changed in a recent commit, marking this as resolved.